app/controllers/application_controller/ci_processing.rb

```diff
@@ -497,53 +497,70 @@ def live_migrate_form_fields
 
   def live_migrate_vm
     assert_privileges("instance_live_migrate")
-    @record = VmOrTemplate.find_by_id(params[:id])
+    @record = VmOrTemplate.find_by(:id => params[:id])
     case params[:button]
     when "cancel"
       add_flash(_("Live Migration of %{model} \"%{name}\" was cancelled by the user") % {
-        :model => ui_lookup(:table => "vm_cloud"), :name => @record.name})
+        :model => ui_lookup(:table => "vm_cloud"), :name => @record.name
+      })
       @record = @sb[:action] = nil
     when "submit"
       if @record.supports_live_migrate?
-        if params['auto_select_host'] == 'on'
-          hostname = nil
-        else
-          hostname = params[:destination_host]
-        end
-        block_migration = params[:block_migration]
-        disk_over_commit = params[:disk_over_commit]
-        begin
-          @record.live_migrate(
-            :hostname         => hostname,
-            :block_migration  => block_migration == 'on',
-            :disk_over_commit => disk_over_commit == 'on'
-          )
-          add_flash(_("Live Migrating %{instance} \"%{name}\"") % {
-            :instance => ui_lookup(:table => 'vm_cloud'),
-            :name     => @record.name})
-        rescue => ex
-          add_flash(_("Unable to live migrate %{instance} \"%{name}\": %{details}") % {
-            :instance => ui_lookup(:table => 'vm_cloud'),
-            :name     => @record.name,
-            :details  => get_error_message_from_fog(ex)}, :error)
+        options = {
+          :hostname         => params['auto_select_host'] == 'on' ? nil : params[:destination_host],
+          :block_migration  => params['block_migration']   == 'on',
+          :disk_over_commit => params['disk_over_commit']  == 'on'
+        }
+        task_id = @record.class.live_migrate_queue(session[:userid], @record, options)
+        unless task_id.kind_of?(Integer)
+          add_flash(_("Instance live migration task %{id} failed.") % {:id => task_id.to_s}, :error)
```

Comment:
I'm having trouble understanding the above logic. If `task_id` is not integer, we're gonna render a flash error
telling that a task with id `task_id` failed? What will `task_id` contain at this point? Will it really be the task's id?
Then it would be an integer, right? But at the same time it wouldn't pass the `unless task_id.kind_of?(Integer)` condition.